### PR TITLE
Implements bracket guides

### PR DESCRIPTION
**Pull request by hediet**
Works efficiently on checker.ts:

image

Uses indentation of brackets (min of start/end column):

image

Overrides indent guides if indent guides are enabled:

image

Active indent guide is only highlighted when there are no bracket guides:

image

Bracket guides can overlap badly formatted text (fixing this is a non-trivial follow up feature and this case should not happen often anyway):

image

See #131001.

### CODE REVIEW

File: src/vs/editor/browser/viewParts/indentGuides/indentGuides.css [hunk 1]
```diff
@@ -3,13 +3,6 @@
...
  *--------------------------------------------------------------------------------------------*/
 
-/*
-	Keeping name short for faster parsing.
-	cigr = core ident guides rendering (div)
-*/
-.monaco-editor .lines-content .cigr {
-	position: absolute;
-}
-.monaco-editor .lines-content .cigra {
```
Comment (hediet) [inline src/vs/editor/browser/viewParts/indentGuides/indentGuides.css:13]: I don't really know how to measure it, but I really doubt that these abbreviations improve performance/memory consumption much. Every frame maybe renders at most up to 100 of those.

File: src/vs/editor/common/controller/cursorColumns.ts [hunk 2]
```diff
@@ -0,0 +1,219 @@
+/*---------------------------------------------------------------------------------------------
```
Comment (hediet) [inline src/vs/editor/common/controller/cursorColumns.ts:1]: I had to move this to a separate file due to cyclic module dependencies.

File: src/vs/editor/common/model/bracketPairColorizer/bracketPairColorizer.ts [hunk 3]
```diff
@@ -82,19 +96,35 @@ export class BracketPairColorizer extends Disposable implements DecorationProvid
...
 	onDidChangeDecorations(listener: () => void): IDisposable {
 		return this.didChangeDecorationsEmitter.event(listener);
 	}
+
+	/**
+	 * Returns all bracket pairs that intersect the given range.
+	 * The result is sorted by the start position.
+	*/
+	getBracketPairsInRange(range: Range): BracketPair[] {
+		this.bracketsRequested = true;
```
Comment (hediet) [inline src/vs/editor/common/model/bracketPairColorizer/bracketPairColorizer.ts:124]: The tree is build the first time brackets are requested and stays alive until the model is disposed.

**Review by alexdima — CHANGES_REQUESTED** [review]
Looks good! I would like to see tests, as a lot of the new code is complicated and it is not covered by existing tests. 

I ran into the following case that crashes the editor:

* repeat the line `this is a large file. this is a large file. this is a large file. this is a large file. this is a large file. this is a large file. this is a large file. this is a large file. this is a large file. this is a large file.` 400k times
* enable `"editor.bracketPairGuides.enabled": true,`
* observe NPE that causes the editor to not render completely

image

File: src/vs/editor/common/controller/cursorCommon.ts [hunk 4]
```diff
@@ -17,7 +15,7 @@ import { LanguageIdentifier } from 'vs/editor/common/modes';
 import { AutoClosingPairs, IAutoClosingPair } from 'vs/editor/common/modes/languageConfiguration';
 import { LanguageConfigurationRegistry } from 'vs/editor/common/modes/languageConfigurationRegistry';
 import { ICoordinatesConverter } from 'vs/editor/common/viewModel/viewModel';
-import { Constants } from 'vs/base/common/uint';
+export { CursorColumns } from './cursorColumns';
```
[the remark this answers is not in the document]
Comment (hediet) [reply src/vs/editor/common/controller/cursorCommon.ts:18]: It has 88 references and these changes would dominate the other changes in this PR. I suggest to do the refactoring separately.

File: src/vs/editor/common/model/textModel.ts [hunk 5]
```diff
@@ -3076,6 +3078,82 @@ export class TextModel extends Disposable implements model.ITextModel, IDecorati
 		return { startLineNumber, endLineNumber, indent };
 	}
 
+	public getLinesBracketGuides(startLineNumber: number, endLineNumber: number, activePosition: IPosition | null, highlightActiveGuides: boolean, includeNonActiveGuides: boolean): model.IndentGuide[][] {
+		const result: model.IndentGuide[][] = [];
+
+		const bracketPairs = this._bracketPairColorizer.getBracketPairsInRange(
```
[the remark this answers is not in the document]
Comment (hediet) [reply src/vs/editor/common/model/textModel.ts:3084]: Bracket guides & colorization behave the same. For extra large file, no tree is constructed in both cases.

File: src/vs/editor/common/model/textModel.ts [hunk 6]
```diff
@@ -3076,6 +3078,82 @@ export class TextModel extends Disposable implements model.ITextModel, IDecorati
...
+		if (activePosition && bracketPairs.length > 0) {
+			const bracketsContainingActivePosition =
+				(startLineNumber <= activePosition.lineNumber && activePosition.lineNumber <= endLineNumber)
+					// Does active position intersect with the view port? -> Intersect bracket pairs with activePosition
+					? bracketPairs.filter(bp => bp.range.containsPosition(activePosition))
+					: this._bracketPairColorizer.getBracketPairsInRange(
+						Range.fromPositions(activePosition)
+					);
+
+			activeBracketPairRange = tail(bracketsContainingActivePosition)?.range;
```
[the remark this answers is not in the document]
Comment (hediet) [reply src/vs/editor/common/model/textModel.ts:3103]: But there also cases like this:

```ts
function test() {
	if (true) { | }
}
```

I think highlighting the range from line 1 to 3 (i.e. effectively line 2) in this case could also be confusing.

I'm open to implement your suggestion though.

File: src/vs/editor/common/viewModel/splitLinesCollection.ts [hunk 7]
```diff
@@ -640,7 +641,135 @@ export class SplitLinesCollection implements IViewModelLinesCollection {
...
+		return this.lines[viewLineInfo.modelLineNumber - 1].getViewLineMinColumn(
+			this.model,
+			viewLineInfo.modelLineNumber,
+			viewLineInfo.modelLineWrappedLineIdx
+		);
+	}
+
+	private getModelStartPositionOfViewLine(viewLineInfo: ViewLineInfo): Position {
+		const line = this.lines[viewLineInfo.modelLineNumber - 1];
+		const minViewColumn = line.getViewLineMaxColumn(
```
[the remark this answers is not in the document]
Comment (hediet) [reply src/vs/editor/common/viewModel/splitLinesCollection.ts:664]: Good catch, this was a copy paste error!

**Review by alexdima — APPROVED** [review]

**Pull request merged**